File: processes/Linux/subproc.c
#include <stdio.h>
#include <stdlib.h>

#define ARGC 2
#define FILENAME_SIZE 32

size_t read_chunk_file(char* filename);
void write_result(size_t result, char* filename);

int main(int argc, char* argv[])
{
    /* check the number of passed arguments */
    if (argc != ARGC) {
        printf("Usage: %s <input_file>\n", argv[0]);
        return EXIT_FAILURE;
    }
    char* filename = argv[1];
    size_t counter = read_chunk_file(filename);

    write_result(counter, filename);

    return 0;
}

size_t read_chunk_file(char* filename)
{
    size_t counter = 0;
    FILE* chunk_file;
    chunk_file = fopen(filename, "r");
    if (!chunk_file) {
        perror("File opening failed");
        exit(EXIT_FAILURE);
    }
    int c;
    // count how many ascii 0x00-0x1F in chunk file
    while ((c = fgetc(chunk_file)) != EOF) {
        if ((c >= 0 && c <= 0x1F) || (c == 0x7F)) {
            ++counter;
        }
    }
    if (ferror(chunk_file)) {
        printf("I/O error when reading\n");
        fclose(chunk_file);
        chunk_file = NULL;
        exit(EXIT_FAILURE);
    }
    else if (feof(chunk_file)) {
        printf("End of file is reached successfully\n");
    }

    fclose(chunk_file);
    chunk_file = NULL;
    return counter;
}

void write_result(size_t result, char* filename)
{
    char* result_filename = malloc(FILENAME_SIZE);
    sprintf(result_filename, "%s_result", filename);
    FILE* output_file = fopen(result_filename, "w");
    if (!output_file) {
        perror("File opening failed");
        exit(EXIT_FAILURE);
    }
    fprintf(output_file, "%d", result);
    // free allocated memory
    fclose(output_file);
    output_file = NULL;
}

File: processes/Python/subproc.py
import sys

def read_chunk_file(filename: str) -> int:
    # read file and count desired ASCII symbols
    counter = 0
    try:
        with open(filename, 'r') as chunk_file:
            while True:
                c = chunk_file.read(1) # read byte by byte
                if not c:
                    break
                if ((ord(c) >= 0 and ord(c) <= 0x1F) or (ord(c) == 0x7F)):
                    counter += 1
        return counter
    except IOError as e:
       print("Error: ", e, file=sys.stderr)
       sys.exit(1)
    
def write_result(result: int, filename: str) -> None:
    result_filename = f"{filename}_result"
    try:
        with open(result_filename, 'w') as output_file:
            output_file.write(str(result))
    except IOError as e:
       print("Error: ", e, file=sys.stderr)
       sys.exit(1)


def main():
    # check if args are passed
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <input_file>", file=sys.stderr)
        sys.exit(1)

    filename = sys.argv[1]
    # read file and count desired ASCII symbols
    result = read_chunk_file(filename)
    # write result in the output file
    write_result(result, filename)

    sys.exit(0)

if __name__ == "__main__":
    main()